### PR TITLE
Don't throw in GetMetadataReference for an unexpected assembly type

### LINKED ISSUE #40466: Searching Solution Explorer crashes or freezes VS
**Visual Studio Version**:

16.5.0 Preview 2 [29616.133.master]

**Steps to Reproduce**:

1. Clone https://github.com/dotnet/fsharp

2. Run `Build.cmd` (might take a little while)

3. Open solution in latest dogfood 16.5 build

Now search for `language` in Solution Explorer.

**Expected Behavior**:

Filters Solution Explorer to have nodes with `language` in them.

**Actual Behavior**:

Filters, but then either immediately crashes or hangs for a while before crashing.

**User Impact**:

Cannot use Solution Explorer search.

### PR DESCRIPTION
**Pull request by RikkiGibson**
Closes #40466

### CODE REVIEW

**Comment by RikkiGibson** [comment]
PTAL @AlekseyTs

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs [hunk 1]
```diff
@@ -1118,7 +1118,12 @@ public new MetadataReference GetMetadataReference(IAssemblySymbol assemblySymbol
 
         private protected override MetadataReference CommonGetMetadataReference(IAssemblySymbol assemblySymbol)
         {
-            return GetMetadataReference(assemblySymbol.EnsureCSharpSymbolOrNull(nameof(assemblySymbol)));
+            if (assemblySymbol is Symbols.PublicModel.Symbol { UnderlyingSymbol: AssemblySymbol underlyingSymbol })
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs]: >Symbol [](start = 54, length = 6)

I think we can go directly to ```AssemblySymbol``` here and avoid a type check on the ```UnderlyingSymbol```.

File: src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb [hunk 2]
```diff
@@ -1232,7 +1232,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
         End Function
 
         Private Protected Overrides Function CommonGetMetadataReference(assemblySymbol As IAssemblySymbol) As MetadataReference
-            Return GetMetadataReference(assemblySymbol.EnsureVbSymbolOrNothing(Of AssemblySymbol)(NameOf(assemblySymbol)))
+            Dim symbol = TryCast(assemblySymbol, AssemblySymbol)
+            If symbol IsNot Nothing Then
+                Return GetMetadataReference(symbol)
+            End If
+
+            Return Nothing
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb]: >Return Nothing [](start = 12, length = 14)

Does this maintain the original behavior when the input is Nothing?

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs [hunk 3]
```diff
@@ -1118,7 +1118,12 @@ public new MetadataReference GetMetadataReference(IAssemblySymbol assemblySymbol
...
+            {
+                return GetMetadataReference(underlyingSymbol);
+            }
+
+            return null;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs]: >return null; [](start = 12, length = 12)

Does this maintain the original behavior when the input is null?

File: src/Compilers/CSharp/Test/Symbol/Compilation/CompilationAPITests.cs [hunk 4]
```diff
@@ -2682,6 +2682,22 @@ public void ConstructMethodWithNullability()
...
+        {
+            var vbComp = CreateVisualBasicCompilation("", referencedAssemblies: TargetFrameworkUtil.GetReferences(TargetFramework.Standard));
+
+            var comp = CreateCompilation("");
+
+            var assembly = (IAssemblySymbol)vbComp.GetBoundReferenceManager().GetReferencedAssemblies().First().Value;
+            var metadataRef = comp.GetMetadataReference(assembly);
+            Assert.Null(metadataRef);
+
+            metadataRef = comp.GetMetadataReference(vbComp.Assembly);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Compilation/CompilationAPITests.cs]: >comp.GetMetadataReference(vbComp.Assembly); [](start = 26, length = 43)

Consider also testing with ```null``` as the input.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1)

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs (hunk 3 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs]: Yes, will add the test.

---
In reply to: [364426266](https://github.com/dotnet/roslyn/pull/40794#discussion_r364426266) [](ancestors = 364426266)

**Comment by RikkiGibson** [comment]
Actually putting this in 16.5-preview2 before it goes to QB mode on Friday.

**Comment by jasonmalinowski** [comment]
Tagging @dotnet/roslyn-ide as a heads up -- there might be other dupes of this floating around where other features might have been broken too. Hopefully those bugs can all be closed.

**Comment by jasonmalinowski** [comment]
@RikkiGibson one question about the fix: is this a case where we're passing something that we shouldn't be (and the failure mode got worse) or this was breaking even in expected inputs?

**Comment by AlekseyTs** [comment]
> is this a case where we're passing something that we shouldn't be (and the failure mode got worse) or this was breaking even in expected inputs?

I would say the first, Assemblysymbol and the Compilation did not belong to the same language (VB vs. C#).
